bootstrap/bootstrap.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if len(snaps) == 0 {
+		return fmt.Errorf("internal error: cannot find os/kernel snap")
+	}
+	for _, fullname := range snaps {
+		bootvar := ""
+		bootvar2 := ""
+
+		// detect type
+		snapFile, err := snap.Open(fullname)
+		if err != nil {
+			return fmt.Errorf("can not read %v", fullname)
+		}
+		// read .sideinfo
+		var si snap.SideInfo
+		siFn := fullname + ".sideinfo"
+		if osutil.FileExists(siFn) {
+			j, err := ioutil.ReadFile(siFn)
+			if err != nil {
+				return err
+			}
+			if err := json.Unmarshal(j, &si); err != nil {
+				return fmt.Errorf("cannot read metadata: %s %s\n", siFn, err)
+			}
+		}
+		info, err := snap.ReadInfoFromSnapFile(snapFile, &si)
+		if err != nil {
+			return fmt.Errorf("can not get info for %v", fullname)
+		}
+		// local install
+		if info.Revision.Unset() {
+			info.Revision = snap.R(-1)
+		}
+
+		switch info.Type {
+		case snap.TypeOS:
+			bootvar = "snappy_os"
+			bootvar2 = "snappy_good_os"
+		case snap.TypeKernel:
+			bootvar = "snappy_kernel"
+			bootvar2 = "snappy_good_kernel"
+			if err := extractKernelAssets(fullname, info); err != nil {
+				return err
+			}
+		}
+
+		name := filepath.Base(fullname)
+		for _, b := range []string{bootvar, bootvar2} {
+			if b != "" {
+				if err := bootloader.SetBootVar(b, name); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func runCommand(cmdStr ...string) error {
+	cmd := exec.Command(cmdStr[0], cmdStr[1:]...)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("cannot run %q: %s (%s)", cmd, err, output)
+	}
+	return nil
+}
+
+func extractKernelAssets(snapPath string, info *snap.Info) error {
+	// FIXME: hrm, hrm, we need to be root for this - alternatively
+	//        we could make boot.ExtractKernelAssets() work on
+	//        a plain .snap file again (i.e. revert bee59a2)
+	if err := os.MkdirAll(info.MountDir(), 0755); err != nil {
+		return err
+	}
+	defer os.Remove(filepath.Dir(info.MountDir()))
+	defer os.Remove(info.MountDir())
+
+	if err := runCommand("mount", snapPath, info.MountDir()); err != nil {
+		return err
+	}
+	defer runCommand("umount", info.MountDir())
+
+	pb := progress.NewTextProgress()
+	if err := boot.ExtractKernelAssets(info, pb); err != nil {
+		return err
+	}
+	return nil
+}
+
+func copyLocalSnapFile(snapName string) error {
+	snapFile, err := snap.Open(snapName)
+	if err != nil {
+		return err
+	}
+	info, err := snap.ReadInfoFromSnapFile(snapFile, nil)
+	if err != nil {
+		return err
+	}
+	// local snap gets sideloaded revision
+	if info.Revision.Unset() {
+		info.Revision = snap.R(-1)
+	}
+	dst := info.MountFile()
+
+	return osutil.CopyFile(snapName, dst, 0)
+}
+
+type downloadOptions struct {
+	Series       string
+	TargetDir    string
+	StoreID      string
+	Channel      string
+	Architecture string
+}
+
+func downloadSnapWithSideInfo(name string, opts *downloadOptions) (string, error) {
```

Comment:
some of this code should actually live in snapstate in the end because it's the symmetrical/needs to co-evolve with InstallPathWithSideInfo